pupil_src/shared_modules/calibration_routines/calibrate_2d.py

```diff
@@ -0,0 +1,264 @@
+    :return: Tuple. First entry is a Boolean, indicating whether the fit was successful. Second entry is a tuple
+    containing the parameters and signature of the fitted polynomials (to be used to initialize MappingFunction2D).
+    """
+
+    calibration_data = np.asarray(calibration_data)
+
+    if binocular:
+        dof_multiplier = 2
+    else:
+        dof_multiplier = 1
+
+    dof = (calibration_data.shape[1] - 2) // dof_multiplier
+    n_params = dof_multiplier * ((degree + 1) ** dof - len(ignored_terms))
+
+    success_x, params_x = optimize_parameters(
+        n_params,
+        calibration_data[:, :-2],
+        calibration_data[:, -2],
+        binocular,
+        dof=dof,
+        degree=degree,
+        ignored_terms=ignored_terms,
+        regularization=regularization,
+        loss_scale=loss_scale,
+    )
+
+    success_y, params_y = optimize_parameters(
+        n_params,
+        calibration_data[:, :-2],
+        calibration_data[:, -1],
+        binocular,
+        dof=dof,
+        degree=degree,
+        ignored_terms=ignored_terms,
+        regularization=regularization,
+        loss_scale=loss_scale,
+    )
+
+    return (success_x and success_y), (params_x, params_y, dof, degree, binocular)
+
+
+def optimize_parameters(
+    n_params,
+    pupil_positions,
+    targets,
+    binocular,
+    dof=2,
+    degree=2,
+    ignored_terms=(),
+    regularization=0.0,
+    loss_scale=-1,
+):
+
+    initial_guess = np.zeros(n_params)
+
+    if (0, 0) not in ignored_terms:
+        initial_guess[0] = np.mean(targets)
+    if binocular:
+        initial_guess[n_params // 2] = initial_guess[0]
+
+    fit_result = scipy.optimize.minimize(
+        calibration_cost,
+        initial_guess,
+        args=(
+            pupil_positions,
+            targets,
+            binocular,
+            dof,
+            degree,
+            ignored_terms,
+            regularization,
+            loss_scale,
+        ),
+    )
+
+    final_params = list(
+        extend_params(fit_result.x, degree, ignored_terms, 0, binocular)
+    )
+
+    return fit_result.success, final_params
+
+
+def calibration_cost(
+    params,
+    pupil_positions,
+    targets,
+    binocular,
+    dof=2,
+    degree=2,
+    ignored_terms=(),
+    regularization=0.0,
+    loss_scale=-1,
+):
+    extended_params = extend_params(params, degree, ignored_terms, 0, binocular)
+
+    if binocular:
+        polynomial_ = PolynomialBinocular(extended_params, dof, degree)
+    else:
+        polynomial_ = PolynomialMonocular(extended_params, dof, degree)
+
+    squared_residuals = (
+        polynomial_([column for column in pupil_positions.T]) - targets
+    ) ** 2
+
+    if loss_scale > 0:
+        cost = 1 / 2 * np.sum(cauchy_loss(squared_residuals, loss_scale))
+    else:
+        cost = 1 / 2 * np.sum(squared_residuals)
+
+    cost += regularization * np.sum(params ** 2)
+
+    return cost
+
+
+def cauchy_loss(residuals, loss_scale):
```

Comment:
Python supports private methods to some extend. If you prepend a function name with `_`, e.g. `def _foo(): ...`, than this method is marked private (actually more similar to protected). Such methods are e.g. not imported if you import the parent module. Functions and methods which are not supposed to be used from the outside should be marked private to reduce the apparent complexity of your module/class. Please prepend your functions with `_` appropriately.

Reply:
I like the overall structure of your functions and classes. They are both short and serve a clear purpose! :+1: